### PR TITLE
Syncing project settings and global settings in offline mode

### PR DESCRIPTION
**Pull request by kpalway**

### CODE REVIEW

File: src/backend/compiler/compiler.cc [hunk 1]
```diff
@@ -1099,27 +1099,124 @@ extern "C" int seashell_preprocessor_run(struct seashell_preprocessor* preproces
...
+  path.insert(path.begin(), preprocessor->question_dir);
+  path.insert(path.begin(), preprocessor->project_dir);
+
+  // attempt question folder .c file
+  file.pop_back();
+  file.push_back("c");
+  path.pop_back();
+  path.push_back(join(file, '.'));
+  std::string attempt = join(path, '/');
   if(std::ifstream(attempt)) return attempt;
```
Comment (e45lee) [inline src/backend/compiler/compiler.cc]: Probably want to use stat(2) here?  http://man7.org/linux/man-pages/man2/stat.2.html

File: src/frontend/js/storage-service.js [hunk 2]
```diff
@@ -277,9 +293,19 @@
...
+            self.deleteProject(project, true);
+          });
+          updatedProjects.forEach(function(project) {
+            self.updateProject(project);
           });
+          if(settings) {
+            self.saveSettings(JSON.parse(settings));
+          }
           self.database.changelog.clear();
+          Dexie.currentTransaction.on('abort', function(ev) {
```
Comment (e45lee) [inline src/frontend/js/storage-service.js]: Pull this to the start of the transaction?

File: lib/libssh2 [hunk 3]
```diff
@@ -1 +1 @@
-Subproject commit 6d553a7bb94966466c4db0abf9a26660e786a275
+Subproject commit 88366b5ec26c1ac16981e81201cabfb3a6bff8b1
```
Comment (e45lee) [inline lib/libssh2]: I think you can resolve this by doing `git checkout  libssh2-1.7.0` from lib/libssh2.

**Comment by e45lee** [comment]
Looks fine, although the segfault is worrisome.  Can you try to fix it?

File: lib/libssh2 (hunk 3 above)
Comment (kpalway) [inline lib/libssh2]: Highest version it allows me to checkout here is 1.5.0

**Comment by kpalway** [comment]
The segfault is troubling, but I can't replicate it when I run the tests or do any manual testing. Can somebody pull from this and see if it occurs for them?

**Comment by kpalway** [comment]
Nevermind, the latest commit should fix the problem.

**Comment by e45lee** [comment]
K. Try a 'git fetch' from the libssh2 first to update it? Otherwise, it
looks good to me, and I'll probably merge it tomorrow morning.

**Comment by kpalway** [comment]
Had to change the remote url for libssh2 to the github one, that seemed to fix the problem.

**Pull request merged**